### ISSUE TITLE
Gradle toolchain fails on M1 macs

### ISSUE THREAD

**Issue by aleksanderlech** [gradle/gradle#19874]
Hello,

Our company is fully "gradlified" and we are now having trouble with the new M1 Macs and support for the Gradle Java Toolchains.

Basically for most of the images the gradle will fail to auto-download arm64 compatible JDK and fail as follows:

```
Starting a Gradle Daemon, 2 incompatible Daemons could not be reused, use --status for details

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':lib:compileJava'.
> Could not resolve all dependencies for configuration ':lib:compileClasspath'.
   > Failed to calculate the value of task ':lib:compileJava' property 'javaCompiler'.
      > Unable to download toolchain matching these requirements: {languageVersion=16, vendor=any, implementation=vendor-specific}
         > Unable to download toolchain. This might indicate that the combination (version, architecture, release/early access, ...) for the requested JDK is not available.
            > Could not read 'https://api.adoptopenjdk.net/v3/binary/latest/16/ga/mac/aarch64/jdk/hotspot/normal/adoptopenjdk' as it does not exist.

```

### Expected Behavior
- Try to fallback to amd64 architecture when M1 is not available
- Consider support for AZUL builds that seem to cover M1 pretty well

### Current Behavior
- Builds breaks on trying to auto-download JDK that is not present in the target repository.

### Context
- Compiling Java projects

### Steps to Reproduce 
- Using M1 based device create the following project:

```
plugins {
    id 'java-library'
}

repositories {
    mavenCentral()
}

java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(16))
    }
}

dependencies {}
```

- execute `./gradlew build`

### Your Environment

System: MacOS 12.1 (21C52)
CPU: Apple M1 Max
Gradle: 7.4
Build scan URL: https://scans.gradle.com/s/nc6ef5nzu4tx4

**Comment by aleksanderlech** [comment]
> This is the plan from gradle folks [#19140 (comment)](https://github.com/gradle/gradle/issues/19140#issuecomment-981457699)
> 
> The ability to provide custom implementations for downloading a jdk will definitely unblock this

Hello @rpalcolea , than it makes this issue a duplicate. Thank you!
However it would be nice to have possibility of implementing own filter criteria or at least have more fallbacks - for instance use other compatible architecture or even try newer JDK already installed in the machine.
